pymc3/distributions/simulator.py

```diff
@@ -14,127 +14,319 @@
+
+    fn = None
+    distance = gaussian
+    sum_stat = identity
+
+    def __new__(cls, *args, **kwargs):
+        if cls.fn is None:
+            raise ValueError("SimulatorRV fn was not specified")
+
+        if cls.ndim_supp is None:
+            raise ValueError(
+                "SimulatorRV must specify `ndim_supp`. This is the minimum"
+                "number of dimensions for a single random draw from the simulator."
+                "\nFor a univariate simulator `ndims_supp = 0`"
+            )
+
+        if cls.ndims_params is None:
+            raise ValueError(
+                "Simulator RV must specify `ndims_params`. This is the minimum"
+                "number of dimensions for every parameter that the Simulator takes,"
+                "including epsilon.\nFor a Simulator that can take two scalar "
+                "parameters` ndims_params = [0, 0, 0]"
+            )
+
+        distance = cls.distance
+        if not isinstance(distance, Op):
+            if distance == "gaussian":
+                distance = gaussian
+            elif distance == "laplace":
+                distance = laplace
+            elif distance == "kullback_leibler":
+                raise NotImplementedError("KL not refactored yet")
+                # TODO: Wrap KL in aesara OP
+                # distance = KullbackLiebler(observed)
+                # if sum_stat != "identity":
+                #     _log.info(f"Automatically setting sum_stat to identity as expected by {distance}")
+                #     sum_stat = "identity"
+            elif callable(distance):
+                distance = create_distance_op_from_fn(distance)
+            else:
+                raise ValueError(f"The distance metric {distance} is not implemented")
+
+        sum_stat = cls.sum_stat
+        if not isinstance(sum_stat, Op):
+            if sum_stat == "identity":
+                sum_stat = identity
+            elif sum_stat == "sort":
+                sum_stat = at.sort
+            elif sum_stat == "mean":
+                sum_stat = at.mean
+            elif sum_stat == "median":
+                # Missing in Aesara, see aesara/issues/525
+                sum_stat = create_sum_stat_op_from_fn(np.median)
+            elif callable(sum_stat):
+                sum_stat = create_sum_stat_op_from_fn(sum_stat)
+            else:
+                raise ValueError(f"The summary statistic {sum_stat} is not implemented")
+
+        cls.distance = distance
+        cls.sum_stat = sum_stat
+        return super().__new__(cls)
+
+    @classmethod
+    def rng_fn(cls, *args, **kwargs):
+        *args, epsilon, size = args
+        return cls.fn(*args, size, **kwargs)
+
+    @classmethod
+    def _distance(cls, epsilon, value, simulated_value):
+        return cls.distance(epsilon, value, simulated_value)
+
+    @classmethod
+    def _sum_stat(cls, value):
+        return cls.sum_stat(value)
+
+
 class Simulator(NoDistribution):
```

Comment:
nitpick, if we are going to have a `Simulator` and a `SimulatorRV`, maybe the first one should be  renamed to `PseudoLikelihood`, `SimulatedLikelihood` `AbcLikelihood` or something similar. One counterargument to this proposal, is that if we are going to distinguish between simulator and pseudolikelihood, the distance and summary statistics should be part of the later not the former.

Reply:
Yeah, that was one of the ideas, the `SimulatorRV` would simply return random draws and the `Pseudolikelihood` would add the `logp` factor. 

The downside is that we don't yet know how to create a "dynamic" `logp` using the optional user defined `sum_stat` and `distance` functions. It means we might need to have users subclass not only the `SimulatorRV` but also the `Pseudolikelihood` if they want to use non-default functions. If someone figures out #4831, then we could simply copy their strategy.